### ISSUE TITLE
Importing runtime support only for option-selected features.

### ISSUE THREAD

**Issue by arv** [google/traceur-compiler#1628]
We get a lot of jabs from some other projects because we do not inline our runtime. We used to do this (like CoffeeScript) but due to demand of supporting single file to single file compiling we moved the code out to an external runtime.

However, we should make it an option to embed the runtime either as an import or by prepending.

For import we could do something like:

```js
class C {}
```

If `transformOptions.classes` is true we would transform this to:

```js
import $traceurRuntime_createClass from 'traceur-runtime:createClass';
class C {}
```

and then in the ClassTransformer we would use `$tracuerRuntime_createClass` instead of `$traceurRuntime.createClass`.

This would need to be done before the module transformer so that the module transformer can handle the import.

The above is of course only valid when we compile `Module`. For `Script` the only option is to prepend the source code.

**Comment by arv** [comment]
Some questions:

What should the module name be? How can we make that work?

For `Script` we need to continue to use `$traceurRuntime.createClass` but we can output a list of runtime functions we need and then we can create the minimal runtime.

Does it work to have modules for runtime?

How can we share the code between runtime modules and runtime scripts? Currently the runtime code is written as `Script` that adds methods to `$traceurRuntime`.

I started hacking on this in a branch: https://github.com/arv/traceur-compiler/tree/inline-runtime-as-modules

**Comment by arv** [comment]
@johnjbarton @domenic @guybedford

**Comment by johnjbarton** [comment]
I was imagining something different with "inline runtime": an implementation that generates inline code rather than calling functions for eg createClass.

The solution outlined here sounds more like "internal import", that is the attachment of dependencies to the transformers so the minimal runtime is created by import rather than by prepending all of the runtime. Also valuable, just a different thing.

**Comment by arv** [comment]
Good point. Inlining the runtime in place would be bad for code size.

The goal is mostly to allow the compiled code to not have to manually add runtime.js

**Comment by johnjbarton** [comment]
The boot strap problem comes down to the module format for the module-store-initialization. For example, if one wants to call `System.registerModule()` when storing modules and `System.get()` when retrieving them, then obviously those two functions and the code they rely on cannot be implemented with modules.  These initial functions must be Script compiled or compiled to eg inline modules.   I don't think anything else needs to be in Script. So I guess it otherwise works for runtime to be modules.

As for the naming, we already have a solution that works with traceur@ for our non-runtime modules.

Generating `import` statements is not essential, and that is where the requirement that module transforms precedes any transforms using runtime. We may not want to force that constraint.

**Comment by johnjbarton** [comment]
The module dependencies are extracted during the parse phase, before any transform is applied. So maybe another approach: 
  1. Refactor `FromOptionsTransformer` to allow derived classes to override the body of the function appended by `append`. 
  2. Implement `TransformsFromOptionsTransformer` to do the current work.
  3. Implement `RuntimeFromOptionsTransformer` to call static method `runtimeModules` on each option-enabled transform; returns the same thing as `getModuleSpecifiers()`.
  4. Run the `RuntimeFromOptionsTransformer` as the first transform or in `InternalLoader` as part of the dependency work.

**Comment by arv** [comment]
My branch currently does 4. It adds a new transformer as the first transformer.

**Comment by arv** [comment]
@guybedford Hi!

The thing that I got stuck on was the name resolution.

**Comment by guybedford** [comment]
@arv can you clarify what you mean by name resolution? I would suggest using simple path resolution - `import 'traceur-runtime/createClass.js'`, which will correspond to a folder containing modules in most environments.

**Comment by arv** [comment]
That would work if someone set up a mapping for 'traceur-runtime'. However, no one has done that by default. To me it is not clear how traceur self hosting would be able to do this. I think @johnjbarton has some ideas but I didn't fully grok what he meant.

**Comment by guybedford** [comment]
@arv, it would be nice to have a custom name configurable for the runtime (`{ runtime: 'traceur-runtime' }`). Perhaps the default for CommonJS would be `require('traceur/runtime/class')` so that it works in NodeJS and Browserify for CommonJS without any config. For the Traceur single-file output, it could just call it `@traceur-runtime/class` and include them in the bundle by default.

**Comment by johnjbarton** [comment]
I think we just need a bootstrap level of support for `traceur-runtime` mapping, similar to the way we support lookup for `@traceur` for testing: https://github.com/google/traceur-compiler/blob/master/src/runtime/ModuleStore.js#L295

**Comment by guybedford** [comment]
@johnjbarton it would be nice to offer a custom resolution option for users not doing single-file bundles.

**Comment by johnjbarton** [comment]
@guybedford you want the output name to be customized? Seems like the map can do that.

**Comment by guybedford** [comment]
@johnjbarton yes that can work too, provided it runs through map.

**Comment by arv** [comment]
Making the name configurable seems plausible but lets focus on getting something basic working first.

**Comment by johnjbarton** [comment]
The bootstrap issues can be solved independent of the features: define a runtime.js that imports modules and get that build.

**Comment by johnjbarton** [comment]
Of course there is a snag....`ModuleStore` was implemented with classes.

**Comment by pflannery** [comment]
Be great if the runtime is generated based on what features are used.
I wonder of it can be determined whats needed to be included and then generate an in memory dynamic module along the lines of:
```js
// bundled-runtime.js
import {core} from 'runtime/core.js'; // compiler path to runtime
import {classes} from 'runtime/classes.js';
import {template} from 'runtime/template.js';
import {properTailCalls} from 'runtime/properTailCalls.js';

// initialization methods go here
properTailCalls.init();

export const $traceurRuntime = {
  core,
  classes,
  template,
  propertTailCalls
  etc....
};
```
Then based on the cli option (see below) this is pre-appended to each module as an import statement (in memory) similar to `import {$traceurRuntime} from './bundled-runtime.js';`
Each runtime module will mainly be written in ES5 except they can use import and export statements because they themselves don't require the $traceurRuntime module.

The cli options needed would be something like `runtime=inline|bundle|none(currently the default now)`
`bundle` would mean it would generate `bundled-runtime.js` file instead of concatenating the runtime in with the single file output.

**Comment by arv** [comment]
@pflannery That is close to what I envisioned. The thing that got me stuck was how to resolve the module name.

```js
// some/dir/test.js
import {createClass} from '????/runtime/classes.js';
let C = createClass(...);
```

But as a temporary step to get closer we can (and should make) a `bundled-runtime.js` which assembles the different runtime modules.

**Comment by pflannery** [comment]
How about:

- generate a temp folder relative to cwd()
- copy the runtime files from the traceur cli path so we end up with something like `import {createClass} from './.traceur/runtime/classes.js';`
- transform the users app to a tree
- detect runtime features required
- generate the dynamic runtime tree
- prepend runtime tree with app tree

**Comment by guybedford** [comment]
In terms of the module name for runtime, the point is that the environment the code runs in needs to be an environment that has a `runtime` package installed into the path output there.

This can be done with one or two options:
* `modularRuntime: true`
* `modularRuntimePath: 'path/to/traceur-runtime' (defaulting to `traceur-runtime` perhaps). I would probably set this to be `traceur/bin/traceur-runtime.js` myself.

If doing single-file builds, there could be an option for:
* `includeRuntimeModule`

In which case, the `System.register('traceur-runtime/some/used/thing')` get added into the single-file bundle as well.

It would be really great to see this.

**Comment by mkeremguc** [comment]
'auto inline modular runtime feature(s) as per need' option for compiler would be very handy and useful
